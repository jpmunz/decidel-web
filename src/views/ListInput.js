import React from "react";
import { Form, InputGroup } from "react-bootstrap";
import "./ListInput.scss";
import RemoveButton from "./RemoveButton";
import autoBind from "react-autobind";

class ListInput extends React.Component {
  constructor(props) {
    super(props);

    const numInitialOptions = this.props.numInitialOptions || 1;
    const initialOptions = [];

    for (let i = 0; i < numInitialOptions; i++) {
      initialOptions.push("");
    }

    this.state = { options: initialOptions };

    autoBind(this);
  }

  handleRemoveOption(i, event) {
    event.preventDefault();
    const newOptions = this.state.options.slice(0);
    newOptions.splice(i, 1);
    this.setState({ options: newOptions });
    this.props.onChange(newOptions);
  }

  handleChangeOption(i, event) {
    const newOptions = this.state.options.slice(0);
    newOptions[i] = event.target.value;
    this.setState({ options: newOptions });
    this.props.onChange(newOptions);
  }

  addOption(value) {
    const newOptions = this.state.options.concat(value);
    this.setState({
      stealFocus: true,
      options: newOptions
    });
    this.props.onChange(newOptions);
  }

  handleAddOption(event) {
    this.addOption(event.target.value);
  }

  render() {
    return (
      <ul className="ListInput">
        {this.state.options.map((option, i) => {
          return (
            <li key={i}>
              <InputGroup>
                <Form.Control
                  data-testid="list-input-option"
                  disabled={this.props.disabled}
                  autoFocus={
                    this.state.stealFocus && i === this.state.options.length - 1
                  }
                  type="text"
                  onChange={event => this.handleChangeOption(i, event)}
                  value={option}
                />

                <InputGroup.Append className="remove-button-container">
                  <RemoveButton
                    tabIndex="-1"
                    disabled={this.props.disabled}
                    onClick={event => this.handleRemoveOption(i, event)}
                  />
                </InputGroup.Append>
              </InputGroup>
            </li>
          );
        })}
        <li>
          <Form.Control
            className="add-another"
            alt="add-another"
            disabled={this.props.disabled}
            type="text"
            value=""
            placeholder={this.props.addMessage}
            onChange={this.handleAddOption}
          />
        </li>
      </ul>
    );
  }
}

export default ListInput;
