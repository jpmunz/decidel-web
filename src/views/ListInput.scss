@import "mixins";

.ListInput {
  li:not(:first-child) {
    padding-top: 20px;
  }

  .remove-button-container {
    padding: 0 10px;
  }
}
