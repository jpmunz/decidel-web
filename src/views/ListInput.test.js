import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ListInput from "./ListInput";

const onChangeMock = jest.fn();

const renderComponent = function({
  numInitialOptions = null,
  disabled = false,
  addMessage = "add new"
}) {
  return render(
    <ListInput
      numInitialOptions={numInitialOptions}
      disabled={disabled}
      addMessage={addMessage}
      onChange={onChangeMock}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Disabling input", () => renderTest({ disabled: true }));
test("Setting more initial options", () =>
  renderTest({ numInitialOptions: 2 }));

test("Adding and removing options", () => {
  const component = renderComponent({ numInitialOptions: 1 });

  fireEvent.change(component.getAllByTestId("list-input-option")[0], {
    target: { value: "Option 1" }
  });

  expect(onChangeMock.mock.calls).toEqual([[["Option 1"]]]);

  fireEvent.change(component.getByAltText("add-another"), {
    target: { value: "New Option" }
  });

  expect(onChangeMock.mock.calls).toEqual([
    [["Option 1"]],
    [["Option 1", "New Option"]]
  ]);

  fireEvent.click(component.getAllByAltText("remove")[0]);

  expect(onChangeMock.mock.calls).toEqual([
    [["Option 1"]],
    [["Option 1", "New Option"]],
    [["New Option"]]
  ]);
});
