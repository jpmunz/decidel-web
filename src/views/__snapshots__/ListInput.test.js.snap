// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Disabling input 1`] = `
<DocumentFragment>
  <ul
    class="ListInput"
  >
    <li>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          data-testid="list-input-option"
          disabled=""
          type="text"
          value=""
        />
        <div
          class="remove-button-container input-group-append"
        >
          <button
            class="ImageButton btn btn-primary"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <img
              alt="remove"
              src="remove.svg"
            />
          </button>
        </div>
      </div>
    </li>
    <li>
      <input
        alt="add-another"
        class="add-another form-control"
        disabled=""
        placeholder="add new"
        type="text"
        value=""
      />
    </li>
  </ul>
</DocumentFragment>
`;

exports[`Setting more initial options 1`] = `
<DocumentFragment>
  <ul
    class="ListInput"
  >
    <li>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          data-testid="list-input-option"
          type="text"
          value=""
        />
        <div
          class="remove-button-container input-group-append"
        >
          <button
            class="ImageButton btn btn-primary"
            tabindex="-1"
            type="button"
          >
            <img
              alt="remove"
              src="remove.svg"
            />
          </button>
        </div>
      </div>
    </li>
    <li>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          data-testid="list-input-option"
          type="text"
          value=""
        />
        <div
          class="remove-button-container input-group-append"
        >
          <button
            class="ImageButton btn btn-primary"
            tabindex="-1"
            type="button"
          >
            <img
              alt="remove"
              src="remove.svg"
            />
          </button>
        </div>
      </div>
    </li>
    <li>
      <input
        alt="add-another"
        class="add-another form-control"
        placeholder="add new"
        type="text"
        value=""
      />
    </li>
  </ul>
</DocumentFragment>
`;
